python/dolfinx/la/petsc.py

```diff
@@ -63,6 +49,79 @@ def create_vector_wrap(x: Vector) -> PETSc.Vec:  # type: ignore[name-defined]
     )
 
 
+def create_vector(
+    maps: typing.Sequence[tuple[IndexMap, int]], kind: typing.Optional[str] = None
+) -> PETSc.Vec:
+    """Create a PETSc vector from a sequence of maps and blocksizes.
+
+    Three cases are supported:
+
+    1. If ``maps=[(im_0, bs_0), ..., (im_n, bs_n)]`` is a sequence of indexmaps and blocksizes and
+       ``kind`` is ``None``or is ``PETSc.Vec.Type.MPI``, a ghosted PETSc vector whith block
+       structure described by ``(im_i, bs_i)`` is created.
+       The created vector ``b`` is initialized such that on each MPI process ``b = [b_0, b_1, ...,
+       b_n, b_0g, b_1g, ..., b_ng]``, where ``b_i`` are the entries associated with the 'owned'
+       degrees-of-freedom for ``(im_i, bs_i)`` and ``b_ig`` are the 'unowned' (ghost) entries.
+
+       If more than one tuple is supplied, the returned vector has an attribute ``_blocks``
+       that holds the local offsets into ``b`` for the (i) owned and
+       (ii) ghost entries for each ``V[i]``. It can be accessed by
+       ``b.getAttr("_blocks")``. The offsets can be used to get views
+       into ``b`` for blocks, e.g.::
+
+           >>> offsets0, offsets1, = b.getAttr("_blocks")
+           >>> offsets0
+           (0, 12, 28)
+           >>> offsets1
+           (28, 32, 35)
+           >>> b0_owned = b.array[offsets0[0]:offsets0[1]]
+           >>> b0_ghost = b.array[offsets1[0]:offsets1[1]]
+           >>> b1_owned = b.array[offsets0[1]:offsets0[2]]
+           >>> b1_ghost = b.array[offsets1[1]:offsets1[2]]
+
+    2. If ``V=[(im_0, bs_0), ..., (im_n, bs_n)]`` is a sequence of function space and ``kind`` is
+       ``PETSc.Vec.Type.NEST``, a PETSc nested vector (a 'nest' of ghosted PETSc vectors) is
+       created.
+
+    Args:
+        map: Sequence of tuples of ``IndexMap`` and the associated block size.
+        kind: PETSc vector type (``VecType``) to create.
+
+    Returns:
+        A PETSc vector with the prescribed layout. The vector is not initialised to zero.
+    """
+    if len(maps) == 1:
+        # Single space case
+        index_map, bs = maps[0]
+        ghosts = index_map.ghosts.astype(PETSc.IntType)  # type: ignore[attr-defined]
+        size = (index_map.size_local * bs, index_map.size_global * bs)
+        return PETSc.Vec().createGhost(ghosts, size=size, bsize=bs, comm=index_map.comm)  # type: ignore
+
+    if kind is None or kind == PETSc.Vec.Type.MPI:
```

Comment:
Can this part be factored out in a standalone function? It's basically repeated verbatim in `dolfinx.fem.petsc._assign_block_data`, and it would be good to have a function from `dolfinx.la` to call from there